### ISSUE TITLE
InAppBrowser popup does not close in case of Facebook Login but works fine for twitter

### ISSUE THREAD

**Issue by davinderkumar** [meteor/meteor#8296]
I am developing an app which has Facebook login and Twitter login options. I am using accounts-facebook and accounts-twitter plugins.

Facebook login works fine on website but InAppBrowser popup does not close after successful authentication in mobile app version. 

Whereas Twitter integration works fine using InAppBrowser popup. 

I tried to use loginStyle: redirect in app but it does not work as OAuth implementation has loginStyle hardcoded to popup in case of cordova apps.

Here is HTML after successful authentication:

```
<body>
  <p id="completedText" style="display:none;">
    Login completed. <a href="#" id="loginCompleted">
      Click here</a> to close this window.
  </p>
  <div id="config" style="display:none;">{"setCredentialToken":false,"storagePrefix":"Meteor.oauth.credentialSecret-","isCordova":true}</div>
  <script type="text/javascript" src="/packages/oauth/end_of_popup_response.js"></script>
</body>
```

but it isn't visible as display is none.

Meteor 1.4.2.3
Here is version info of each package:
accounts-facebook                    1.0.11  Login service for Facebook accounts
accounts-twitter                     1.1.12  Login service for Twitter accounts

I tested it on 
- iPhone 7 plus with iOS 10.2
- Simulator with iOS 10.2
- Simulator with iOS 9.3 
Same results on all of these.

I tried to search all over internet for a solution but could not find anything. Please provide a solution.

**Comment by abernix** [comment]
If you go into the Developer Tools of your browser on on that popup, is there an error listed?

**Comment by abernix** [comment]
Can you package your problem up into an easy to run `meteor`-app, upload it to GitHub and provide a link to it?  Someone can provide their own Facebook app credentials, but the key here is to have an easy reproduction.

**Comment by davinderkumar** [comment]
I have attached barebones meteor app to reproduce the issue. Here is the link:
[fbtest.zip](https://github.com/meteor/meteor/files/758610/fbtest.zip)
Please change APP_ID and APP_SECRET in main.js in server folder.

This sample works fine in web version but has issue in cordova version.

In local environment, I saw this error in server console: 
> W20170207-23:49:40.582(5.5) (oauth.js:428) Error in OAuth Server: Failed to complete OAuth handshake with Facebook. failed [400] {"error":{"message":"Error validating verification code. Please make sure your redirect_uri is identical to the one you used in the OAuth dialog request","type":"OAuthException","code":100,"fbtrace_id":"CFY7yM5g5UR"}}

It cannot be fixed as the redirect_uri works only with redirect loginStyle but loginStyle value is overridden with "popup" for Cordova apps.  Refer line: 50-52, oauth.js

Thanks,

**Comment by davinderkumar** [comment]
@abernix were you able to look into this issue?

**Comment by davinderkumar** [comment]
@abernix any luck?

**Comment by Denly** [comment]
Same here. My Cordova app shows blank after I logged in by FB. I inspected it. It's displays: none 
![screen shot 2017-02-15 at 4 01 33 pm](https://cloud.githubusercontent.com/assets/7588279/22995269/3711395a-f398-11e6-9408-2991e133cfe1.png)

**Comment by Denly** [comment]
btw @davinderkumar I asked in stackoverflow. Waiting for solution.  http://stackoverflow.com/questions/42261166/fb-oauth-failed-while-log-in-meteor-cordova-app

**Comment by santteegt** [comment]
Hi all. I could fix the error **Error validating verification code. Please make sure your redirect_uri is identical to the one you used in the OAuth dialog request** by setting the _$ROOT_URL_ environment variable with the redirect URL value set in the Facebook Login settings. WIth that I'm still getting the same blank UIWebView without closing behaviour but, at least, when I tap the Done button, the user is logged in on my app.

I'm going to keep working on a solution for this, as I'm just a few days ahead the launch of our app. If someone finds a solution for this, please let us know. I'll appreciate it.

Cheers

**Comment by trustinbharat** [comment]
Hi

I am having a similar issue on my website. The user is just stuck at the blank screen after Facebook Sign-in. If he presses refresh, he will be shown logged-in but mostly, user leaves thinking there is an error. Can u please help me with the solution.

Regards

**Comment by josephtactica** [comment]
Hi all, any updates on this?

**Comment by josephtactica** [comment]
I got this working.

- Removed and re-added `accounts-facebook` and `service-configuration`.
- `ROOT_URL=http://random.ngrok.io meteor run ios-device --mobile-server=http://random.ngrok.io`

Without setting `ROOT_URL`, Fb gets stuck on a blank screen and won't redirect. Make sure `ROOT_URL` and `--mobile-server` have the same URL. I'm still not sure what caused this...

**Comment by ghybs** [comment]
Hi all,

**TL; DR:** `meteor add cordova:cordova-plugin-inappbrowser@1.7.1`

While `ROOT_URL` and `--mobile-server` indeed seem required in the case of Facebook login, there is still a common issue for _all_ OAuth logins within iOS Cordova packaged apps by Meteor, still present with current 1.5 release.

As OP described, the user is presented with a "new" window (Meteor Cordova apps always use the InAppBrowser plugin, irrespective of the `loginStyle` option) that displays the 3rd party OAuth service.
After successful authentication and permissions authorization, in iOS the window becomes blank and remains as-is, the Meteor app does not take control back, and user has to manually touch the "DONE" blue button on bottom left for that window to close and the Meteor app to re-appear.
![image](https://user-images.githubusercontent.com/12232803/27332251-279345da-55d2-11e7-8b6a-001677d46613.png)
This is very confusing for users to remain "stuck" on a white page. They may think there is an error with the login or the app. Actually the login did succeed, only the window closing failed.

This is not reproducible on Android or non-Cordova packaged apps.

This issue has been specifically raised in [[CB-11136]](https://issues.apache.org/jira/browse/CB-11136):

> InAppBrowser fails to close with WKWebView OAuth
>
> Launching InAppBrowser from Cordova iOS Platform 4+ with its WKWebView for OAuth (e.g. Facebook or Google login) fails to close as it should.
>
> The reason is that the entire WKWebView thread seems to pause when another view controller is presented. This can be confirmed by inspecting the WKWebView session in Safari, running `window.open('http://something.com')` and then trying to enter another command into that Safari console.

This has been fixed as of plugin version [`1.7.0`](https://github.com/apache/cordova-plugin-inappbrowser/blob/master/RELEASENOTES.md#170-feb-28-2017).

Unfortunately Meteor `1.6-beta.2` is [still using](https://github.com/meteor/meteor/blob/release/METEOR%401.6-beta.2/packages/oauth/package.js#L59-L61) InAppBrowser version `1.6.1`

Manually upgrading to `1.7.0` or later fixes the issue on iOS Cordova packaged apps for all tested OAuth login services.

**Comment by abernix** [comment]
@ghybs Thanks for the detailed information.  I'm currently looking into a Cordova update for an upcoming release (See #8686).  As major updates to plugins will typically require re-publishing the app in the app store, I've marked this so we can get all these together in one update.

**Comment by davinderkumar** [comment]
I'm still getting this error. Here is my version

Meteor 1.5.4.1:

Packages:
service-configuration@1.0.11
iron:router
facebook-config-ui 
dynamic-import
accounts-facebook    --- version 1.2.1

Cordova 
cordova:cordova-plugin-inappbrowser  2.0.1
cordova:cordova-plugin-splashscreen  4.0.3
cordova:cordova-plugin-statusbar     2.2.3

Command used to run :
meteor run ios-device --mobile-server 10.91.0.83:3000

FB App contains these URL in "Client OAuth Settings":
http://10.91.0.83:3000/_oauth/facebook?close 
http://10.91.0.83:3000/

Issue: After successful authentication from App, app stuck on white screen.

Console error on Client:
Error in Success callbackId: InAppBrowser714380123 : SyntaxError: JSON Parse error: Unexpected identifier "_"

Error on server:
Error in OAuth Server: Failed to complete OAuth handshake with Facebook. failed [400] {"error":{"message":"Error validating verification code. Please make sure your redirect_uri is identical to the one you used in the OAuth dialog request","type":"OAuthException","code":100,"fbtrace_id":"A3pzgVblLuD"}}

Web version of app works fine. Only mobile version has problem

**Comment by davinderkumar** [comment]
@abernix @hwillson 
is there any update for this?
I'm getting this issue in 1.6.0.1 as well.
I have tried 
ROOT_URL=http://www.xyz.com meteor run ios-device --settings=settings.json --mobile-server www.xyz.com
and without ROOT_URL as well. 
App gets stuck at white screen.
on local environment ( if I use mobile-server as localhost:3000) , then it works fine.

i'm using cordova:cordova-plugin-inappbrowser 2.0.1

console shows this error
Error in Success callbackId: InAppBrowser714380123 : SyntaxError: JSON Parse error: Unexpected identifier "_"

**Comment by davinderkumar** [comment]
This is working now in 1.6.1.1. I'm not sure about previous version.

Some notes:
Hostname / IP should be same for ROOT_URL, "mobile-server", Facebook App -> Login -> Valid OAuth Redirect URIs

E.g. this worked fine
ROOT_URL=http://10.91.0.83:3000 meteor run ios-device  --mobile-server 10.91.0.83:3000

And Valid OAuth Redirect URIs was 
http://10.91.0.83:3000/_oauth/facebook?close

Here is same App source code which is working fine.

[fbtest  1.6.1.1.zip](https://github.com/meteor/meteor/files/2053748/fbtest.1.6.1.1.zip)

**Comment by arichter83** [comment]
For me using `loginStyle: "redirect"` solved the problem. I used `screen.width` to differentiate, which method to use.
